src/coin-selection/test/Spec.hs

```diff
@@ -392,4 +394,147 @@ matchingIndexMP = do
+  stakeKey <- C.generateSigningKey C.AsStakeKey
+  let
+    withdrawalAmount = 10_000_000
+
+    stakeHash =
+      C.verificationKeyHash . C.getVerificationKey $ stakeKey
+
+    stakeCred = C.StakeCredentialByKey stakeHash
+
+    stakeCert =
+      C.makeStakeAddressRegistrationCertificate
+      . C.StakeAddrRegistrationPreConway C.ShelleyToBabbageEraBabbage
+      $ stakeCred
+
+    delegationCert =
+      C.makeStakeAddressDelegationCertificate
+      $ C.StakeDelegationRequirementsPreConway C.ShelleyToBabbageEraBabbage stakeCred poolId
+
+    stakeCertTx = BuildTx.execBuildTx $ do
+      BuildTx.addCertificate stakeCert
+
+    delegCertTx = BuildTx.execBuildTx $ do
+      BuildTx.addCertificate delegationCert
+
+  -- activate stake
+  void $ tryBalanceAndSubmit mempty Wallet.w2 stakeCertTx []
+  -- delegate to pool 
+  void $ tryBalanceAndSubmit mempty Wallet.w2 delegCertTx [C.WitnessStakeKey stakeKey]
+
+  -- modify the ledger state
+  setReward stakeCred (C.quantityToLovelace withdrawalAmount)
+
+  (rewards, delegations) <- queryStakeAddresses (Set.fromList [stakeCred]) Defaults.networkId
+
+  when (length rewards /= 1) $ fail "Expected 1 reward"
+  when (length delegations /= 1) $ fail "Expected 1 delegation"
+
+withdrawalTest :: forall m. (MonadIO m, MonadMockchain m, MonadError BalanceTxError m, MonadFail m) => m ()
+withdrawalTest = withRegisteredPool Wallet.w1 $ \ poolId -> do
+  stakeKey <- C.generateSigningKey C.AsStakeKey
+  let
+    withdrawalAmount = 10_000_000
+
+    stakeHash =
+      C.verificationKeyHash . C.getVerificationKey $ stakeKey
+
+    stakeCred = C.StakeCredentialByKey stakeHash
+
+    stakeCert =
+      C.makeStakeAddressRegistrationCertificate
+      . C.StakeAddrRegistrationPreConway C.ShelleyToBabbageEraBabbage
+      $ stakeCred
+    stakeAddress = C.makeStakeAddress Defaults.networkId stakeCred
+
+    delegationCert =
+      C.makeStakeAddressDelegationCertificate
+      $ C.StakeDelegationRequirementsPreConway C.ShelleyToBabbageEraBabbage stakeCred poolId
+
+    stakeCertTx = BuildTx.execBuildTx $ do
+      BuildTx.addCertificate stakeCert
+
+    delegCertTx = BuildTx.execBuildTx $ do
+      BuildTx.addCertificate delegationCert
+
+    withdrawalTx = execBuildTx $ do
+      BuildTx.addWithdrawal stakeAddress withdrawalAmount (C.KeyWitness C.KeyWitnessForStakeAddr)
+
+  -- activate stake
+  void $ tryBalanceAndSubmit mempty Wallet.w2 stakeCertTx []
+  -- delegate to pool 
+  void $ tryBalanceAndSubmit mempty Wallet.w2 delegCertTx [C.WitnessStakeKey stakeKey]
+
+  -- modify the ledger state
+  setReward stakeCred (C.quantityToLovelace withdrawalAmount)
+
+  -- withdraw rewards
+  void $ tryBalanceAndSubmit mempty Wallet.w2 withdrawalTx [C.WitnessStakeKey stakeKey]
+
+{-| Run the 'Mockchain' action with registered pool
+-} 
+withRegisteredPool :: forall m a. (MonadIO m, MonadMockchain m, MonadError BalanceTxError m, MonadFail m) => Wallet -> (C.PoolId -> m a) -> m a
```

Comment:
The `withXXX :: (a -> IO b) -> IO b`  usually implies that something is done after the action is run. Since that's not the case here we should change `withRegisteredPool` to `registerPool :: (MonadIO m, MonadMockchain m, ...) => Wallet -> m C.PoolId`

Also move to a new module `Convex.MockChain.Staking` in `convex-coin-selection`

Reply:
Yeah this makes more sense.